### PR TITLE
Do not use SQLITE_OPEN_SHAREDCACHE

### LINKED ISSUE #29599: Possible multi-threading issue in Roslyn's sqlite usage
**Version Used**: 
2.9.0-beta4-63001-02

**Steps to Reproduce**:

1. Not sure about the steps to reproduce, it happens seemingly random in VSMac

**Expected Behavior**:

Thread-safety is well defined for SQLite usage.

**Actual Behavior**:
Thread-safety is not well defined in some cases.

Looking at the value of `SQLitePCL.raw.sqlite3_threadsafe ()` advised [here](https://www.sqlite.org/c3ref/threadsafe.html), we get `1` back, which means that we get full thread safety via [Serialized](https://www.sqlite.org/compile.html#threadsafe).

So far so good, that means sqlite will handle synchronization.

The [SqlConnection](http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces.Desktop/Workspace/SQLite/Interop/SqlConnection.cs,56) to the database used is done using `SQLITE_OPEN_NOMUTEX`, which is not the equivalent of the default, `Serialized`, but [Multi-threaded](https://www.sqlite.org/c3ref/open.html), so it does not handle mutexes on the prepared statements.

The locked database is not the same as a busy database, as per [the docs](https://www.sqlite.org/cvstrac/wiki?p=DatabaseIsLocked). This means that we're doing hitting case 3 from that link, which is running 2 selects in parallel with no mutexes.

```
ERROR [2018-08-30 08:32:59Z]: Roslyn error: StorageDatabase_Exceptions Kind=Microsoft.CodeAnalysis.SQLite.Interop.SqlException|Reason=Microsoft.CodeAnalysis.SQLite.Interop.SqlException: database table is locked: SolutionData1
database table is locked
  at Microsoft.CodeAnalysis.SQLite.Interop.SqlConnection.Throw (SQLitePCL.sqlite3 handle, Microsoft.CodeAnalysis.SQLite.Interop.Result result) [0x00000] in /_/src/Workspaces/Core/Desktop/Workspace/SQLite/Interop/SqlConnection.cs:284 
  at Microsoft.CodeAnalysis.SQLite.Interop.SqlConnection.ThrowIfNotOk (SQLitePCL.sqlite3 handle, Microsoft.CodeAnalysis.SQLite.Interop.Result result) [0x00003] in /_/src/Workspaces/Core/Desktop/Workspace/SQLite/Interop/SqlConnection.cs:276 
  at Microsoft.CodeAnalysis.SQLite.Interop.SqlConnection.ThrowIfNotOk (Microsoft.CodeAnalysis.SQLite.Interop.Result result) [0x00000] in /_/src/Workspaces/Core/Desktop/Workspace/SQLite/Interop/SqlConnection.cs:270 
  at Microsoft.CodeAnalysis.SQLite.Interop.SqlConnection.ThrowIfNotOk (System.Int32 result) [0x00000] in /_/src/Workspaces/Core/Desktop/Workspace/SQLite/Interop/SqlConnection.cs:267 
  at Microsoft.CodeAnalysis.SQLite.Interop.SqlStatement.Reset () [0x00000] in /_/src/Workspaces/Core/Desktop/Workspace/SQLite/Interop/SqlStatement.cs:43 
  at Microsoft.CodeAnalysis.SQLite.Interop.ResettableSqlStatement.Dispose () [0x00000] in /_/src/Workspaces/Core/Desktop/Workspace/SQLite/Interop/ResettableSqlStatement.cs:34 
  at Microsoft.CodeAnalysis.SQLite.SQLitePersistentStorage+Accessor`3[TKey,TWriteQueueKey,TDatabaseId].TryGetRowIdWorker (Microsoft.CodeAnalysis.SQLite.Interop.SqlConnection connection, TDatabaseId dataId, System.Int64& rowId) [0x00047] in /_/src/Workspaces/Core/Desktop/Workspace/SQLite/SQLitePersistentStorage.Accessor.cs:209 
  at Microsoft.CodeAnalysis.SQLite.SQLitePersistentStorage+Accessor`3[TKey,TWriteQueueKey,TDatabaseId].TryGetRowId (Microsoft.CodeAnalysis.SQLite.Interop.SqlConnection connection, TDatabaseId dataId, System.Int64& rowId) [0x00000] in /_/src/Workspaces/Core/Desktop/Workspace/SQLite/SQLitePersistentStorage.Accessor.cs:183 
  at Microsoft.CodeAnalysis.SQLite.SQLitePersistentStorage+Accessor`3[TKey,TWriteQueueKey,TDatabaseId].ReadBlob (Microsoft.CodeAnalysis.SQLite.Interop.SqlConnection connection, TDatabaseId dataId) [0x00000] in /_/src/Workspaces/Core/Desktop/Workspace/SQLite/SQLitePersistentStorage.Accessor.cs:142 
  at Microsoft.CodeAnalysis.SQLite.SQLitePersistentStorage+Accessor`3+<ReadStreamAsync>d__11[TKey,TWriteQueueKey,TDatabaseId].MoveNext () [0x000f6] in /_/src/Workspaces/Core/Desktop/Workspace/SQLite/SQLitePersistentStorage.Accessor.cs:82 
```

### PR DESCRIPTION
**Pull request by sharwell**
This flag can only be used when the application manages concurrency for database write operations to ensure no more than one connection is writing data at any given point. Roslyn only provides the weaker
guarantee that each connection is only used from one thread at a time, but may be used for reading or writing.

Fixes #29599

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Out of curiosity, why not go the readerwriterlock route?

**Comment by sharwell** [comment]
> Out of curiosity, why not go the readerwriterlock route?

Many/most of the readers would need to be upgradeable, which is equivalent to just locking everything.

**Comment by CyrusNajmabadi** [comment]
> Many/most of the readers would need to be upgradeable, which is equivalent to just locking everything.

Hrmm... that seems odd to me.  You'd only need to upgrade to a write lock if there were actually writes to process.  For lots of other things (like reading blobs) you'd only need the read lock.

**Comment by heejaechang** [comment]
can we get some perf data to see the impact and have test to make sure it doesn't happen again?

**Comment by sharwell** [comment]
> can we get some perf data to see the impact and have test to make sure it doesn't happen again?

We have no known perf scenarios or observable repro cases for this bug.

**Comment by heejaechang** [comment]
but this is foundamental service according to @CyrusNajmabadi  and according to you, changing founamental service, we need to make sure perf doesn't get affected.

since RPS won't create sqlite db for picasso scenario, we need manual test to make sure it didn't regress.

and we need tests to make sure on such foundamental service's behavior change and make sure it actually fix the issue.

**Comment by sharwell** [comment]
Memory usage will get impacted. The additional caches will consume up to 512KiB\*(*n*-1), where *n* is the maximum number of thread pool threads executing code concurrently at any point in the application life.

**Comment by CyrusNajmabadi** [comment]
> since RPS won't create sqlite db for picasso scenario, we need manual test to make sure it didn't regress.

We did the RPS tests in these cases originally (by forcing sqlite to always be on).  And that was in its original form which did not have this flag.  Things passed fine then (outside of the extra DLL loads).  Reverting back to the form that already passed all these tests nd which doesn't have the current regression seems acceptable.

**Comment by heejaechang** [comment]
then run the RPS test again and provide info so other people can see?

**Comment by heejaechang** [comment]
we should have discussed perf diff between esent and sqlite before we made complete change. 

1 thing I know is that I used to save checksum for each sourcetext in a solution in esent and re-read it without calculating checksum again next time the same solution is opened and that was faster than calculating all source text checksum with esent. but with sqlite it is way slower.

calculating all checksum for sourcetext in roslyn solution, usually takes about 5 seconds on my machine. with esent cache, it took about 2 seconds. with sqlite, it was more than 10 seconds.

actually, I am not the one who discovered this perf degrade, it was @CyrusNajmabadi who came and complained OOP is so slow. at the end, I removed the cache since cache made it slower than just calculating checksums.

it could be sqlite being just slower than esent, or it could be the way it interacts with sqlite (bytes array, sqlite, over pooled stream, esent) makes too much garbage. I dont know. 

or at the time, sqlite was shared with multiple processes (every VS, OOP that havesame solution opened), so it could be from that.

but I do know new persistent service made this particular case way slower.

and sure I dont have data. so it will be good to start have data on it.

**Comment by heejaechang** [comment]
not sure why asking me to collect perf data. I am asking collect perf data for this PR so we can actual data to show it doesn't affect perf much.

**Comment by heejaechang** [comment]
since the orignal implementation, 

we had 2 issues we know. one is it caused VS to OOM which @sharwell  fixed since sqlite created a lot connections that consumes giga bytes of memory.

we moved to WAL mode since sqlite was too slow. and that made can't remember exact number but quite a bit of improvement on perf.

so, I really hardly believe original implementation of on pal with esent.

**Comment by CyrusNajmabadi** [comment]
> we had 2 issues we know. one is it caused VS to OOM which @sharwell fixed since sqlite created a lot connections that consumes giga bytes of memory.

To be clear This wasn't sqlite doing this. :)  This was an issue with the code i wrote where i allowed an async call while someone was holding onto a connection.  Because there was no guarantee on when that async call might return (and we had some other bugs that led to threadpool explosion), we effectively kept handing out connections to more and more threadpool threads.  The simple fix there was to hold/release the connection before/after the async point, thus ensuring that we only ever had a small outstanding number of pooled connections.

> And that made can't remember exact number but quite a bit of improvement on perf.

We need to understand what's going on.  Literally all that is happening are batched writes of key/value pairs.  Literally the bread and butter for this system.  Original perf testing showed scalability of this far beyond what we could hope to throw at it.  I'm worried we updated to some sort of version/lib that regressed this, or we're hitting some sort of pathological case.  

I woudl highly recommend that we actually use the real tools for measuring performance here instead of guessing.  For example, original testing used the built-in timer and trace functionality fo sqlite to get accurate measurements from teh DB itself to prove out what's going on.  This is what was done with the original impl to get real timings and to know the impact of any change.  

if necessary, we should just include the code in the product now (ifdef'ed out potentially) so that explicit measurements can be easily taken as we adjust things.

**Comment by heejaechang** [comment]
sure I didn't mean the sqlite bits or dll itself. I mean the whole system we replaced whether it is the sqlite or our side that uses it. 

anyway, so I think we are in agreement then we need to perf data before proceeding.

**Comment by CyrusNajmabadi** [comment]
> so, I really hardly believe original implementation of on pal with esent.

It was given the absence of issues elsewhere.  For example, the connection-async problem stemmed from something else stalling out the threadpool.  So the async work we were kicking off in between things wasn't actually a problem during measuring *until* we ran into that other problem.  That other problem revealed this implementation issue (which wasn't difficult to rectify), but it wasn't something you would experience normally.

So, definitely, tehre can be issues and bugs that come up.  But there is nothing intrinsic about this system that makes it any slower than esent.  It's not like sqlite was actually slow here.  It's just that, in certain circumstances, our wrapper just kept on responding to requests, and there was an allocation overhead to that that was addressed.

**Comment by heejaechang** [comment]
> ut there is nothing intrinsic about this system that makes it any slower than esent. 

provide data,.

**Comment by CyrusNajmabadi** [comment]
>But there is nothing intrinsic about this system that makes it any slower than esent.

The data was provided with teh original work.  I no longer have it as it is on machines that i returned to MS.   I have no ability to do anything related to RPS or use any of the MS perf testing infrastructure.  I cannot provide such things.

**Comment by heejaechang** [comment]
alright, then let's create new perf data so people can have refresehed data set with current implementation. and add tests that prevent this kind of thing happens again.

I am not sure why you arguing about these request. these are ones you guys always ask to others?

**Comment by CyrusNajmabadi** [comment]
> I am not sure why you arguing about these request. 

I'm not sure why you think i'm arguing :)  I'm providing data and information here.  

>  these are ones you guys always ask to others?

I think you've been misreading my posts.  As i've said on numerous posts, i agree with you.  For example:

>> and sure I dont have data. so it will be good to start have data on it.
> Agreed. I would definitely like to know what is causing the poor behavior.

**Comment by heejaechang** [comment]
alright then! misunderstood!

**Comment by heejaechang** [comment]
@sharwell 
> The additional caches will consume up to 512KiB*(n-1), where n is the maximum number of thread pool threads executing code concurrently at any point in the application life

so each thread consuming 512K is only VM? how about allocations? so as thread pool created and go away will it create new 512K alloctaions and throw away? from our history of caring about even boxing allocations, is that okay?

and can you point me where we set 512K as cache for each sql connections? do we have any data saying that 512K is actually what is happening?

**Comment by CyrusNajmabadi** [comment]
> so as thread pool created and go away will it create new 512K alloctaions and throw away?

No, the connection objects are pooled and handed out to incoming threads on demand.  So say, you have 8 max threadpool threads (though individual threads may come and go), we'd only have a max of 8 potential connectoin objects to hand out to them.

We are weak to the runaway threadpool problem.  i.e. if something else in hte system makes it so that threadpool starts behaving badly (i.e. starvation) and it keeps spawning new thread that call in here, then we'll make more and more connection objects which will be added to the pool.

We may want to cap the total number of connections we'll pool at some reasonable about (maybe 2x core count?).  That way, if the rest of the system starts going bad (i.e. with threadpool starvation) that doesn't cause any problems here.

However, if the threadpool is behaving well, we shouldn't have more than N pooled connections, with one pooled connection per threadpool thread-count.

**Comment by sharwell** [comment]
> so each thread consuming 512K is only VM? how about allocations?

It's a native memory allocation for the connection cache. Managed allocations are not impacted by this cache.

> so as thread pool created and go away will it create new 512K alloctaions and throw away?

The allocations which occur are permanent until the application closes. The upper bound on connection count equals the largest count of live threads in the thread pool, and would only be achieved in a situation where all thread pool threads are operating on an `SqlConnection` concurrently.

> and can you point me where we set 512K as cache for each sql connections? do we have any data saying that 512K is actually what is happening?

https://github.com/dotnet/roslyn/issues/22650#issue-264405238

It's actually 478.125KiB per connection, but it's easier to reason about 512KiB IMO.

**Comment by CyrusNajmabadi** [comment]
Oh, i even mentioned capping the number of pooled connections here: https://github.com/dotnet/roslyn/issues/22650#issuecomment-335638013

This could be done fairly trivially over in SQLitePersistentStorage.ReleaseConnection.  Just check how big the pool is, and close and do not keep around the underlying connection object if it's past a certain point (16 connections would likely be more than sufficient for any reasonable scenario).

**Comment by heejaechang** [comment]
I suggested capped as well which I think @sharwell rejected. can't remember what the reason was.

by the way, so do we have data that this concurrency actually help perf? rather than let SQLite handle them? mutax, serialization or anything else? what was bring us to build up our pool?

**Comment by CyrusNajmabadi** [comment]
> by the way, so do we have data that this concurrency actually help perf? rather than let SQLite handle them?

I'm not sure what that means per-se.  To read/write anything from sqlite you need a connection.  Making the connections is not super expensive, but it's not free either.  Each time you do it, sqlite actually goes and connects to the file, reads in all appropriate metadata about it and forms its in-memory representation of things like the DB schema.

Also, by keeping around connections we can do things like construct sqlite statements (the things that are actually executed) and cache those for the lifetime of the connection.  For example, we will be doing a bazillion reads and writes, all of the form:

```$@"select rowid from ""DocumentData"" where ""DataId"" = ?"```   or
```$@"insert or replace into "DocumentData}""(""DataId"",""Data"") values (?,?)"```

Without a pooled connection, each time we did that sqlite would have to parse the above, make a prepared statement, insert the relevant parts into the ?'s, come up with a plan, and execute it.  Caching these means that work is only ever done once, and all we're doing from that point on is saying: just do the select, or the insert/replace again, only with these stubbed in values for those ? locations

Basically, you can totally get rid of connections, it just means that every single operation we do takes the hit of all of these every time.

Now, we could use one connection.  However, then we would have every single read/write blocking on this single connection (as sqlite doesn't allow multiple threads to use a single connection object concurrently).  With pooled connections, each thread can have their own connection, and things like reads can safely concurrently overlap.

**Comment by heejaechang** [comment]
thank you for the explanation. 

so it is not actually cache per thread, it is concurrent connected connections. with async, 1 thread could have 10 connections opened then? 

and what you are saying is multiple connections improve reading by support concurrent read but write is still serialized?

and you created those flush thing to improve write performance?

I believe you said we are write heavy not read heavy. have we measured what write perf we should expect?

**Comment by CyrusNajmabadi** [comment]
> so it is not actually cache per thread, it is concurrent connected connections. with async, 1 thread could have 10 connections opened then?

There is no code in the impl that should cause this to happen.  The way we've written it is that any thread that wants to talk to the DB will grab a pooled connection, then synchronously talk, then release it.  It is considered "not allowed" to do async work while you have grabbed a pooled connection instance.

This is exactly the issue that @sharwell fixed.  Note that nothing enforces this. But the number of places we get connections is so small, and hte work we do is so trivial, it's easy to audit and ensure that people do not do this when making changes here.

**Comment by heejaechang** [comment]
so what you are saying is, since PooledConnection is private to Sqlite Persistent Service and not exposed to outside. it is okay just having comment don't do it is all we need? sure, i would still put upper cap though. especially it is only guarded by code review.

**Comment by CyrusNajmabadi** [comment]
> and what you are saying is multiple connections improve reading by support concurrent read but write is still serialized?

Individual writes are serialized against each other yes.  Reads can happen concurrent with writes thanks to the WAL.  See 1.2 here: https://www.sqlite.org/wal.html

> WAL provides more concurrency as readers do not block writers and a writer does not block readers. Reading and writing can proceed concurrently. 

Because writers can block other writers, we have a single queue of writes that need to happen.  When people want to write they just add to that queue and immediately return (so writing is effectively free to the calling thread).  

When reads come along, we want to make sure they see any 'completed' writes.  However, because reads are only to a specific table/row, we only write out any pending writes to that specific table/row combo.

i.e. we may have 10k outstanding write requests (for example, some sort of index for all the docs in a solution).  However, a read for a table/row coming in will explicitly force the write for that table/row and then do the read.

--

Note: this does mean the worst case scenario is something producing data, writing it, and then immediately reading it in some sort of loop.  However, when i analyzed our features nothing behaved in this way.  Features either read things in and used it, or computed data and wrote it (but then used the computed data).  Features that wrote out data to free up memory did so in large swaths and hten read in large swaths of data.  They didn't alternate between small writes/reads/writes/reads etc.

**Comment by CyrusNajmabadi** [comment]
> I believe you said we are write heavy not read heavy. have we measured what write perf we should expect?

Effectively db writing should be nearly inconsequential compared with the cost of actually producing the data to write out.  i.e. our index computation consumes a bunch of CPU, while the writing of that data is just a handful of IOps twice a second.

**Comment by heejaechang** [comment]
and that implementation only works since we don't support a transaction. alright, good to know. since it uses different pooled connections in one Read or Write, it doesn't guarantee timing or transaction. alright, since our persistent service was not used for such case so far. bad that we can never do so but, sure.

**Comment by CyrusNajmabadi** [comment]
> so what you are saying is, since PooledConnection is private to Sqlite Persistent Service and not exposed to outside. it is okay just having comment don't do it is all we need?

Yes.  And i agree, it would be good to comment here.  PooledConnection and SqlConnection shoudl both have docs that say: once you get an instance of this, you should do your work as quickly as possible and then release it.  You should definitely *not* do anything async while holding either of these.

Thanks for pointing this out.

> sure, i would still put upper cap though. especially it is only guarded by code review.

I agree.  The upper cap is still useful in the case that some *other* part of the system has caused runaway threadpool thread growth.  it prevents us from pooling an unbounded number of connections.  This is def not desirable as we'll still be holding onto all those connections, even once the threadpool restores itself to normalcy.  i.e. say something else causes threadpool starvation and we get 1000 threads (it happens...).  That could lead to 1000 pooled sqlite connections.  Even once the threadpool gets down to a normal thread count, we'd still have 1k pooled connections, even though we'd only ever then hand out a handful of them.

**Comment by heejaechang** [comment]
so how big our write is usually? and how often do we write? I know 4 things we use the service. diagnostics, todo comments, syntax tree index and symbol tree index.

what is the size of each stream we have to sqlite? and how long those usually takes? 

I would interest to know how long read usually takes as well for those data.

**Comment by CyrusNajmabadi** [comment]
> and that implementation only works since we don't support a transaction. 

Wait... can you clarify what you mean by that?  We definitely support transactions (and we use them) :)

**Comment by CyrusNajmabadi** [comment]
>  since it uses different pooled connections in one Read or Write, it doesn't guarantee timing or transaction.

It does guarantee than any fully completed write to a table/row will be seen by the next read of that table/row (as long as there isn't some other write to the same table/row).


In other words, if you simultaneously issued (pseudocode) Read(table, row) and Write(table, row) it's non deterministic what would happen.  However, if you did:  `await Write(table, row); Read(table, row)` then you'd see the table from the write, even if the Read was performed on a different thread.

This is because Reads of a Table/Row force writes *of only that table/row* to be processed before issuing the read.

**Comment by CyrusNajmabadi** [comment]
> so how big our write is usually? and how often do we write? I know 4 things we use the service. diagnostics, todo comments, syntax tree index and symbol tree index.

I have forgotten that precise information unfortunately.  It was small enough that writing out 4MB twice a second was more than enough headroom to handle things.  in other words, the amount of writes we produced was far less than 8MB/s.   I know i wanted to give us a *lot* of headroom.  i.e. i was concerned about someone being on a spinny HDD (i.e. not on an ssd) that was very overloaded.  So i wanted to know that we could absorb a lot of write requests during that time, and that we could issue single writes that would dump a lot of data at once .

**Comment by heejaechang** [comment]
> Wait... can you clarify what you mean by that? We definitely support transactions (and we use them) :)

before, I did some experiment feature (not related to persistent service) that internally used esent (not through persisetent layer but used esent as structure data base) with the normal transactions and stuff.

let's say, I want to use sqlite not with persistent service but want to use it directly. current PooledConnection not allowing any async method in it is a bit too much restrictive.

but I guess, one can create new db (new db file) and doesnt use code in sqlite persistenet and create its own sqlite wrapper to use it might work. 

for me, I still used all the esent accessor and stuff, I just didn't use Read/Write that force things to be stream.

**Comment by CyrusNajmabadi** [comment]
> what is the size of each stream we have to sqlite? and how long those usually takes?

The size of stream depends on how many writes that were made.  I don't remember those numbers right now.  It depends on things like sizof(indices) and whatnot.

How long it takes it takes can be broken into two costs.  One normal cost per write, and one amortized cost.

The normal cost is that you take the disk hit to write that data just to disk.  In general this size is basically the size of hte byte[]s you asked the persistent storage to write, plus a small amount of additional metadata from sqlite.  So, if you wanted to write out 100k of byte[]s, that would be roughly 100k * some_small_metadata_overhead.  

That's the cost you get every time we write out a batch of writes to disk.  It is dependent on your HD speed to write this to the WAL.

There is an additional write cost *some* of the time after this.  Specifically, this WriteAheadLog is occasionally sucked back in and absorbed into the main DB file.  IIRC, sqlite does this after every 4MB (1000 x 4k pages) is written to the WAL by default.  So, for something like Roslyn (which is currently at 18MB for me), that is effective 5 writes of the WAL back into the DB file.

**Comment by CyrusNajmabadi** [comment]
> let's say, I want to use sqlite not with persistent service but want to use it directly. current PooledConnection not allowing any async method in it is a bit too much restrictive.

Well, it would be ok.  You would just release the connection if you needed to do async work, an then acquire a new one when you went back.  Basically, for the actual reading/writing you have a connection.  In between, if you need to do other stuff (*especially unbounded async stuff), you just don't hold onto the connection and you allow someone else to use it.  sqlite connections are safe to use by multiple threads.  Just *not at the same time*.  So one thread can get the connection, use it for the actual read/write to the db.  It can then let the connection go (and another thread can then use it), do other work, then reacquire some connection (possibly a different one entirely, or a new one) and use that for any remaining work.

**Comment by heejaechang** [comment]
another reason I was curious was because the uses of byte[]. esent I made it to use Stream that pools byte[] with Stream itself pooled as well. meaning, even saving 4M, 10M stream won't actually allocates too much churn since we pool those 4K byte[].

sqlite seems require 4MB or 8MB bytes[] to write that size data since it doesn't have abstraction over it. so that is why I was curious what the general size of data is since it could create bunch of churn to create those byte[] for data.

**Comment by heejaechang** [comment]
> Well, it would be ok. You would just release the connection if you needed to do async work, an then acquire a new one when you went back.

I dont believe transcation is supported by multiple connections.

**Comment by CyrusNajmabadi** [comment]
> I dont believe transcation is supported by multiple connections.

That is true.  but what is the scenario where you need to open a transaction, do some work, then do some *async* work, then do some more transaction work...?

This would generally be considered a bad idea with *any* DB system as transactions are heavyweight and are not encouraged to be held open longer than necessary.  

Is this just speculative, or do you envision an actual need for that sort of thing?

**Comment by CyrusNajmabadi** [comment]
> I dont think that is true. people who uses Sql server doesn't usually care holding a connection open since a connection is usually cheap. committing is the part that is expensive.

I was speaking specifically about *transactions* not *connections*.   Holding open a transaction is costly.  While that transaction is open the DB system has to make sure that any and all changes made by other connections is not visible to your transaction scope.  That means all sorts of strategies for then how writes are handled to ensure that.  If you then have may people doing this, the costs continue to increase.  

So, it is not wise to start a transaction, do some work with the db, then go away for a while (i.e. for async work), then come back to the db for more work, the to commit.

**Comment by heejaechang** [comment]
>  Holding open a transaction is costly. 

sure, opening transaction is costly since it will hold onto snapshot of when it is opened, and hold onto changes made until commit or rollback. not that's like if it takes hours. not like seconds 

anyway, doesn't matter. one can use sqlite directly when need structure db.

..

SqlPersistentStorage_Helpers.GetBytes

so looked the code, first question is why it is not reusing existing our byte pool
http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Utilities/ObjectPools/SharedPools.cs,60

if it require different size byte[], probably good to put it here so that other feature can share the pooled byte[]

but it is only for small data like when size is small < 4K. not sure where 124K came from.

previously, I think any data less than 4MB reused bytes.

but I also see, it seems limitation of sqlite. it only accepts byte[], length but doesn't support incremental write where one can give bytes[], position, length for same row in a loop.

**Comment by CyrusNajmabadi** [comment]
> so looked the code, first question is why it is not reusing existing our byte pool

The existing byte pool has poor characteristic under concurrency.  It can easily 'lose' objects and not actually pool them.  This actually leads to lots of garbage.  I started there, but quickly moved off that to a non-lossy pool that would ensure no garbage, even in concurrent scenarios.

> but I also see, it seems limitation of sqlite. it only accepts byte[], length but doesn't support incremental write where one can give bytes[], position, length for same row in a loop.

You can do this.  But it's tricker, especially of you need to resize.  It was not a problem in actual measurements so i didn't go down that path.  Effectively, you use this API: https://www.sqlite.org/c3ref/blob_open.html

**Comment by CyrusNajmabadi** [comment]
> sure, opening transaction is costly since it will hold onto snapshot of when it is opened, and hold onto changes made until commit or rollback. not that's like if it takes hours. not like seconds

Right, but hte point is:  if you do an async call in the middle of this, you have no real idea of how long it may take.  If the threadpool is totally occupied (as it naturally can be in VS), you may actually take quite a long time.  So, best to be on the safe side and just not do it at all :)

**Comment by CyrusNajmabadi** [comment]
> previously, I think any data less than 4MB reused bytes.

We could definitely tweak things here if necessary. However, as the numbers above indicate, there's almost no need for this.  Literally 99%+ of all usage goes through the pool.  So unless there is some good data indicating otherwise, i would not really touch anything here.

**Comment by sharwell** [comment]
@heejaechang thanks. I opened a live instance on a system with many cores and it seems to confirm that it's likely to enter a situation where many pooled connections exist (in this case about 140% of the number of logical processors). I agree it's something we'll want to address, especially with things like ThreadRipper 2 getting into customer hands. It won't be uncommon to see developers with 128 logical processors soon and things would be getting out of hand.
